File: backend/app/schemas/technical.py
from __future__ import annotations

from typing import Dict, List, Optional

from pydantic import BaseModel


class IndicatorSignal(BaseModel):
    name: str
    signal: str  # "bullish" | "bearish" | "neutral"
    strength: float
    latest_values: Dict[str, Optional[float]]
    metadata: dict


class TechnicalAnalysisResponse(BaseModel):
    symbol: str
    timeframe: str
    indicators: List[IndicatorSignal]
    overall_signal: str  # "bullish" | "bearish" | "neutral"
    overall_score: float  # -1.0 (極度看空) ~ +1.0 (極度看多)


class SingleIndicatorResponse(BaseModel):
    symbol: str
    timeframe: str
    indicator: IndicatorSignal


# --- 指標時間序列 (圖表覆蓋用) ---


class TimeSeriesPoint(BaseModel):
    time: int
    value: Optional[float]


class IndicatorSeriesData(BaseModel):
    name: str
    lines: Dict[str, List[TimeSeriesPoint]]


class IndicatorSeriesResponse(BaseModel):
    symbol: str
    timeframe: str
    indicator: IndicatorSeriesData


class SupportResistanceLevel(BaseModel):
    price: float
    label: str


class SupportResistanceResponse(BaseModel):
    symbol: str
    timeframe: str
    levels: List[SupportResistanceLevel]

File: frontend/src/lib/types/technical.ts
export type Signal = "bullish" | "bearish" | "neutral";

export interface IndicatorSignal {
  name: string;
  signal: Signal;
  strength: number;
  continuous_score: number;
  latest_values: Record<string, number | null>;
  metadata: Record<string, unknown>;
}

export interface TechnicalAnalysisResponse {
  symbol: string;
  timeframe: string;
  indicators: IndicatorSignal[];
  overall_signal: Signal;
  overall_score: number;
  consensus: number;
}

export interface SingleIndicatorResponse {
  symbol: string;
  timeframe: string;
  indicator: IndicatorSignal;
}

// --- 指標時間序列 (圖表覆蓋用) ---

export interface TimeSeriesPoint {
  time: number;
  value: number | null;
}

export interface IndicatorSeriesData {
  name: string;
  lines: Record<string, TimeSeriesPoint[]>;
}

export interface IndicatorSeriesResponse {
  symbol: string;
  timeframe: string;
  indicator: IndicatorSeriesData;
}

export interface SupportResistanceLevel {
  price: number;
  label: string;
}

export interface SupportResistanceResponse {
  symbol: string;
  timeframe: string;
  levels: SupportResistanceLevel[];
}

export type OverlayIndicator = "ema" | "bbands" | "sr" | "vpvr";
export type SubChartIndicator = "rsi" | "kd" | "macd";
